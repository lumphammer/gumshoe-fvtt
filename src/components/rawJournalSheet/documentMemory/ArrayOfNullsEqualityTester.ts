import { expect } from "vitest";

function isArrayOfNulls(candidate: unknown): candidate is null[] {
  if (!(candidate instanceof Array)) {
    return false;
  }
  for (const value of candidate) {
    if (value !== null) {
      return false;
    }
  }
  return true;
}

export class ArrayOfNulls {
  length: number;
  constructor(length: number) {
    this.length = length;
  }

  equals(other: unknown) {
    return isArrayOfNulls(other) && other.length === this.length;
  }
}

function testIsArrayOfNulls(a: unknown, b: unknown): boolean | undefined {
  if (a instanceof ArrayOfNulls && !(b instanceof ArrayOfNulls)) {
    return a.equals(b);
  } else if (!(a instanceof ArrayOfNulls) && b instanceof ArrayOfNulls) {
    return b.equals(a);
  } else {
    return undefined;
  }
}

expect.addEqualityTesters([testIsArrayOfNulls]);
