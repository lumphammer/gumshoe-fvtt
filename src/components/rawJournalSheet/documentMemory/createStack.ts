import { Stack } from "./types";

export function createStack(periodicity: number): Stack {
  return {
    edits: new Array(periodicity).fill(null),
    bombBay: new Array(periodicity - 1).fill(null),
    next: null,
  };
}
