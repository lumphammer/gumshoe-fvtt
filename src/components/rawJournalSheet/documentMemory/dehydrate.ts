import { BareDocumentMemory, DocumentMemory } from "./types";

/**
 * Convert the working memory into one with the state and snapshots removed
 */
export function dehydrate(documentMemory: DocumentMemory): BareDocumentMemory {
  return {
    stack: documentMemory.stack,
    serial: documentMemory.serial,
    period: documentMemory.period,
    maxDepth: documentMemory.maxDepth,
  };
}
