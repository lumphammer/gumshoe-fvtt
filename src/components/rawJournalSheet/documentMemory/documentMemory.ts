import diff from "textdiff-create";

import { isMagicSerial } from "./isMagicSerial";
import { DocumentMemory, Edit, Stack } from "./types";

function createStack(periodicity: number): Stack {
  return {
    edits: new Array(periodicity).fill(null),
    snapshot: "",
    next: null,
  };
}

export function createHistory(periodicity: number): DocumentMemory {
  return {
    stack: createStack(periodicity),
    serial: 0,
    period: periodicity,
  };
}

function push(
  stack: Stack,
  newState: string,
  serial: number,
  period: number,
  depth: number,
): Stack {
  const newDelta: Edit = {
    change: diff(stack.snapshot, newState),
    timestamp: new Date().toISOString(),
  };
  let next = stack.next;
  if (isMagicSerial(period, depth, serial)) {
    next = push(
      next ?? createStack(period),
      stack.snapshot,
      serial,
      period,
      depth + 1,
    );
  }
  return {
    edits: [...stack.edits.slice(1), newDelta],
    snapshot: newState,
    next,
  };
}

export function save(history: DocumentMemory, state: string): DocumentMemory {
  const serial = history.serial + 1;
  const stack = push(history.stack, state, serial, history.period, 1);
  return {
    stack,
    serial,
    period: history.period,
  };
}
