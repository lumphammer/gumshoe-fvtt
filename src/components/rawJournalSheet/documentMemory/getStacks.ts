import { BareDocumentMemory, Stack } from "./types";

export function getStacks(memory: BareDocumentMemory): Stack[] {
  const stacks: Stack[] = [];
  for (let stack: Stack | null = memory.stack; stack; stack = stack.next) {
    stacks.push(stack);
  }
  return stacks;
}
