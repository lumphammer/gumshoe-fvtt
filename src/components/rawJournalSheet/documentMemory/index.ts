export { createHistory, save } from "./documentMemory";
export { DocumentMemory } from "./types";
