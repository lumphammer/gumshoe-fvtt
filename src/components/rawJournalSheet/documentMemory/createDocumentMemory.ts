import { createStack } from "./createStack";
import { DocumentMemory } from "./types";

/**
 * Create a new empty document memory
 */
export function createDocumentMemory(periodicity: number): DocumentMemory {
  return {
    stack: createStack(periodicity),
    serial: 0,
    period: periodicity,
  };
}
