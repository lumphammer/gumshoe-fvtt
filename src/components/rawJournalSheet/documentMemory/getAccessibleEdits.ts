import { getStacks } from "./getStacks";
import { DocumentMemory, Edit } from "./types";

export function getAccessibleEdits(memory: DocumentMemory) {
  const stacks = getStacks(memory).reverse();
  const edits: Edit[] = stacks.flatMap((stack) => stack.edits);
  return edits;
}
