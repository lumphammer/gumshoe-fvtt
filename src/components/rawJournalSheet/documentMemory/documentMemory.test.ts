import { afterAll, beforeEach, describe, expect, test, vi } from "vitest";

import { ArrayOfNulls } from "./ArrayOfNullsEqualityTester";
import { createHistory, save } from "./documentMemory";

const epoch = "1970-01-01T00:00:01.000Z";

function advanceTime10s() {
  vi.setSystemTime(new Date().setSeconds(new Date().getSeconds() + 10));
}

beforeEach(() => {
  vi.setSystemTime(new Date(epoch));
});

afterAll(() => {
  vi.useRealTimers();
});

describe("createHistory", () => {
  test.each([3, 4, 5, 6, 100, 1000])(
    "creates an empty history with period %i",
    (period) => {
      const history = createHistory(period);
      expect(history).toEqual({
        stack: {
          edits: new ArrayOfNulls(period),
          snapshot: "",
          next: null,
        },
        serial: 0,
        period,
      });
    },
  );
});

test("starts", () => {
  const h1 = createHistory(3);
  const h2 = save(h1, "foo");
  expect(h2).toEqual({
    stack: {
      edits: [
        null,
        null,
        {
          change: [[1, "foo"]],
          timestamp: epoch,
        },
      ],
      snapshot: "foo",
      next: null,
    },
    serial: 1,
    period: 3,
  });
});

test("stores a sequence of edits", () => {
  let h = createHistory(3);
  const states = ["foo", "foobar", "bar"];
  for (const state of states) {
    h = save(h, state);
    advanceTime10s();
  }
  expect(h).toMatchSnapshot();
});

test("stores a sequence of edits onto the next stack", () => {
  let h = createHistory(3);
  const states = ["foo", "foobar", "bar", "barbaz"];
  for (const state of states) {
    h = save(h, state);
    advanceTime10s();
  }
  expect(h).toMatchSnapshot();
});

test("stores a sequence of edits onto the third stack", () => {
  let h = createHistory(3);
  const states = [
    "a",
    "ab",
    "abc",
    "abcd",
    "abcde",
    "abcdef",
    "abcdefg",
    "abcdefgh",
    "abcdefghi",
    "abcdefghij",
  ];
  for (const state of states) {
    h = save(h, state);
    advanceTime10s();
  }
  expect(h).toMatchSnapshot();
});
