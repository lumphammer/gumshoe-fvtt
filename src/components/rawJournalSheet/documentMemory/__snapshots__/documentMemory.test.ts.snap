// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stores a sequence of edits 1`] = `
{
  "period": 3,
  "serial": 3,
  "stack": {
    "edits": [
      {
        "change": [
          [
            1,
            "foo",
          ],
        ],
        "timestamp": "1970-01-01T00:00:01.000Z",
      },
      {
        "change": [
          [
            0,
            3,
          ],
          [
            1,
            "bar",
          ],
        ],
        "timestamp": "1970-01-01T00:00:11.000Z",
      },
      {
        "change": [
          [
            -1,
            3,
          ],
          [
            0,
            3,
          ],
        ],
        "timestamp": "1970-01-01T00:00:21.000Z",
      },
    ],
    "next": null,
    "snapshot": "bar",
  },
}
`;

exports[`stores a sequence of edits onto the next stack 1`] = `
{
  "period": 3,
  "serial": 4,
  "stack": {
    "edits": [
      {
        "change": [
          [
            0,
            3,
          ],
          [
            1,
            "bar",
          ],
        ],
        "timestamp": "1970-01-01T00:00:11.000Z",
      },
      {
        "change": [
          [
            -1,
            3,
          ],
          [
            0,
            3,
          ],
        ],
        "timestamp": "1970-01-01T00:00:21.000Z",
      },
      {
        "change": [
          [
            0,
            3,
          ],
          [
            1,
            "baz",
          ],
        ],
        "timestamp": "1970-01-01T00:00:31.000Z",
      },
    ],
    "next": {
      "edits": [
        null,
        null,
        {
          "change": [
            [
              1,
              "bar",
            ],
          ],
          "timestamp": "1970-01-01T00:00:31.000Z",
        },
      ],
      "next": null,
      "snapshot": "bar",
    },
    "snapshot": "barbaz",
  },
}
`;

exports[`stores a sequence of edits onto the third stack 1`] = `
{
  "period": 3,
  "serial": 10,
  "stack": {
    "edits": [
      {
        "change": [
          [
            0,
            7,
          ],
          [
            1,
            "h",
          ],
        ],
        "timestamp": "1970-01-01T00:01:11.000Z",
      },
      {
        "change": [
          [
            0,
            8,
          ],
          [
            1,
            "i",
          ],
        ],
        "timestamp": "1970-01-01T00:01:21.000Z",
      },
      {
        "change": [
          [
            0,
            9,
          ],
          [
            1,
            "j",
          ],
        ],
        "timestamp": "1970-01-01T00:01:31.000Z",
      },
    ],
    "next": {
      "edits": [
        {
          "change": [
            [
              1,
              "abc",
            ],
          ],
          "timestamp": "1970-01-01T00:00:31.000Z",
        },
        {
          "change": [
            [
              0,
              3,
            ],
            [
              1,
              "def",
            ],
          ],
          "timestamp": "1970-01-01T00:01:01.000Z",
        },
        {
          "change": [
            [
              0,
              6,
            ],
            [
              1,
              "ghi",
            ],
          ],
          "timestamp": "1970-01-01T00:01:31.000Z",
        },
      ],
      "next": null,
      "snapshot": "abcdefghi",
    },
    "snapshot": "abcdefghij",
  },
}
`;
